File: frontend/src/services/cep.ts
/**
 * Serviço para integração com API de CEP
 */

import api from "./api";

export interface AddressData {
  cep: string;
  street: string;
  neighborhood: string;
  city: string;
  state: string;
  complement?: string;
}

export interface CEPResponse {
  success: boolean;
  data?: AddressData;
  error?: string;
}

/**
 * Busca dados de endereço por CEP
 */
export const fetchAddressByCEP = async (cep: string): Promise<CEPResponse> => {
  try {
    // Remove formatação do CEP
    const cleanCEP = cep.replace(/\D/g, "");

    if (cleanCEP.length !== 8) {
      return {
        success: false,
        error: "CEP deve ter 8 dígitos",
      };
    }

    const response = await api.get(`/cep/${cleanCEP}`);

    if (response.data.success) {
      return {
        success: true,
        data: response.data.data,
      };
    } else {
      return {
        success: false,
        error: response.data.error || "Erro ao buscar CEP",
      };
    }
  } catch (error: any) {
    console.error("Erro ao buscar CEP:", error);

    if (error.response?.status === 404) {
      return {
        success: false,
        error: "CEP não encontrado",
      };
    }

    return {
      success: false,
      error: "Erro ao conectar com o serviço de CEP",
    };
  }
};

/**
 * Valida formato do CEP
 */
export const validateCEPFormat = async (
  cep: string,
): Promise<{ isValid: boolean; error?: string }> => {
  try {
    const cleanCEP = cep.replace(/\D/g, "");

    if (cleanCEP.length !== 8) {
      return {
        isValid: false,
        error: "CEP deve ter 8 dígitos",
      };
    }

    const response = await api.get(`/cep/${cleanCEP}/validate`);

    return {
      isValid: response.data.valid,
      error: response.data.valid ? undefined : "CEP inválido",
    };
  } catch (error: any) {
    console.error("Erro ao validar CEP:", error);
    return {
      isValid: false,
      error: "Erro ao validar CEP",
    };
  }
};

File: backend/app/api/v1/cep.py
"""Endpoints para consulta de CEP e autocompletar endereço."""

from fastapi import APIRouter, HTTPException, Depends
from typing import Dict, Any
from app.services.cep_service import CEPService
from app.dependencies import get_current_user
from app.schemas.user_schema import UserInDB
import logging

logger = logging.getLogger(__name__)

router = APIRouter()


@router.get("/cep/{cep}", response_model=Dict[str, Any])
async def get_address_by_cep(
    cep: str, current_user: UserInDB = Depends(get_current_user)
) -> Dict[str, Any]:
    """Busca dados de endereço pelo CEP.

    Args:
        cep: CEP a ser consultado (com ou sem formatação)
        current_user: Usuário autenticado

    Returns:
        Dados do endereço encontrado

    Raises:
        HTTPException: Se CEP não for encontrado ou houver erro na consulta
    """
    try:
        logger.info(f"Usuário {current_user.username} consultando CEP: {cep}")

        address_data = await CEPService.get_address_by_cep(cep)

        if not address_data:
            raise HTTPException(
                status_code=404,
                detail="CEP não encontrado. Verifique se o CEP está correto.",
            )

        logger.info(
            f"CEP {cep} encontrado com sucesso para usuário {current_user.username}"
        )

        return {
            "success": True,
            "data": address_data,
            "message": "Endereço encontrado com sucesso",
        }

    except HTTPException:
        raise

    except Exception as e:
        logger.error(
            f"Erro ao consultar CEP {cep} para usuário {current_user.username}: {str(e)}"
        )
        raise HTTPException(
            status_code=500, detail="Erro interno do servidor ao consultar CEP"
        )


@router.get("/cep/{cep}/validate")
async def validate_cep(
    cep: str, current_user: UserInDB = Depends(get_current_user)
) -> Dict[str, Any]:
    """Valida formato do CEP sem fazer consulta completa.

    Args:
        cep: CEP a ser validado
        current_user: Usuário autenticado

    Returns:
        Resultado da validação
    """
    try:
        clean_cep = CEPService._clean_cep(cep)
        is_valid = CEPService._validate_cep_format(clean_cep)

        return {
            "success": True,
            "data": {
                "cep": cep,
                "clean_cep": clean_cep,
                "is_valid": is_valid,
                "formatted_cep": (
                    CEPService._format_cep(clean_cep) if is_valid else None
                ),
            },
            "message": "CEP válido" if is_valid else "CEP inválido",
        }

    except Exception as e:
        logger.error(f"Erro ao validar CEP {cep}: {str(e)}")
        raise HTTPException(
            status_code=500, detail="Erro interno do servidor ao validar CEP"
        )
